src/components/Forms/PhoneNumberField.js

```diff
@@ -0,0 +1,300 @@
+    }
+
+    if (!this.state.hasValue && nextProps.value) {
+      this.setState({hasValue: true})
+    }
+  }
+
+  getValue = () => {
+    if (!this.input) {
+      return null
+    }
+
+    return this.input.value.replace(/\(|\)|-| /g, '')
+  }
+
+  handleInputChange = (e) => {
+    const { onChange } = this.props
+    const { hasValue } = this.state
+    const value        = e.target.value
+    // Limit setState call to only when hasValue changes
+    if (value && !hasValue) {
+      this.setState({hasValue: true})
+    } else if (!value && hasValue) {
+      this.setState({hasValue: false})
+    }
+
+    onChange && onChange(e, value.replace(/\(|\)|-| /g, ''), value)
+  }
+
+  handleInputFocus = (e) => {
+    this.setState({isFocused: true})
+    this.props.onFocus && this.props.onFocus(e)
+  }
+
+  handleInputBlur = (e) => {
+    this.setState({isFocused: false})
+    this.props.onBlur && this.props.onBlur(e)
+  }
+
+  handleKeyDown = (e) => {
+    this.props.onKeyDown(e)
+  }
+
+  render() {
+    const {
+      floatingLabelText,
+      defaultValue,
+      disabled,
+      fullWidth,
+      halfWidth,
+      hasError,
+      id,
+      isValid,
+      name,
+      required,
+      serverError,
+      validationErrorText,
+      style,
+      value,
+      helperText,
+      autoComplete,
+      snacksTheme
+    } = this.props
+
+    const {
+      hasValue,
+      isFocused
+    } = this.state
+
+    const inputId = id
+
+    return (
+      <div
+        style={[
+          styles.wrapper,
+          fullWidth && styles.fullWidth,
+          halfWidth && styles.halfWidth,
+          style
+        ]}
+      >
+
+        {serverError && !disabled && !isValid &&
+          <ServerError
+            text={serverError}
+          />
+        }
+
+        <div style={styles.inputContainer}>
+          <FloatingLabel
+            text={floatingLabelText}
+            float={isFocused || hasValue}
+            disabled={disabled}
+            isActive={isFocused}
+            hasError={hasError}
+            htmlFor={inputId}
+            style={{pointerEvents: 'none'}}
+          />
+
+          <TextFieldHint
+            text={'(555) 555-555'}
+            show={!hasValue && isFocused}
+            disabled={disabled}
+          />
+
+          <MaskedTextInput
+            mask={['(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/]}
+            id={inputId}
+            guide={false}
+            name={name}
+            aria-required={required}
+            aria-invalid={hasError}
+            onBlur={this.handleInputBlur}
+            onChange={this.handleInputChange}
+            onFocus={this.handleInputFocus}
+            onKeyDown={this.handleKeyDown}
+            autoComplete={autoComplete}
+            placeholder=""
+            defaultValue={value !== undefined ? undefined : defaultValue}
+            disabled={disabled}
+            keepCharPositions={true}
+            render={(ref, props) => (
```

Comment:
The fabled render prop :D

Reply:
In all it's glory. It was super helpful in this situation :D